### PR TITLE
AirfRANS: volume-loss-weight=30x for aggressive volume improvement

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS has a 4.5x gap on volume MSE (0.00764 vs SpiderSolver's 0.0017) while surface MSE already beats SpiderSolver by 9x (0.000482 vs 0.0043). The model is clearly under-weighting volume in the loss function.

PRs #3101 (3x) and #3102 (10x) test moderate volume loss weights. Given the 4.5x gap, even 10x may be insufficient — the model needs to dramatically shift its attention toward volume fields. 30x explicitly tells the optimizer that volume errors are 30x more important than surface errors, which should force the model to reallocate capacity.

The surface MSE has 9x headroom above SpiderSolver — we can afford significant surface regression if volume improves proportionally.

## Instructions

**Run 1: volume-loss-weight=30x**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 2 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --volume-loss-weight 30.0 \
  --wandb_group af-volume-weight-sweep
```

If train.py doesn't have --volume-loss-weight, add it: multiply the volume component of the loss by this factor before combining with the surface loss. The default should be 1.0 (current behavior).

Report BOTH metrics:
- val_primary/surface_mse (baseline: 0.000482, SpiderSolver: 0.0043)
- val_primary/volume_mse (baseline: 0.00764, SpiderSolver: 0.0017)

## Baseline

- **val_primary/surface_mse:** 0.000482 (PR #2951 MERGED, beats SpiderSolver 0.0043 by 9x)
- **val_primary/volume_mse:** 0.00764 (4.5x above SpiderSolver 0.0017)
- **Config:** 2L/256d, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, no-EMA
- **Reproduce:** `cd target/icml2026 && python train.py --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT megumi:

## Results

### Run 1: volume-loss-weight=30x

| Metric | Baseline | This Run (best val) | SpiderSolver | Delta vs Baseline |
|--------|----------|-------------------|--------------|-------------------|
| val_primary/surface_mse | 0.000482 | 0.001264 (ep377) | 0.0043 | +162% (2.6x worse) |
| full_val/volume_mse | 0.00764 | 0.006235 (ep396) | 0.0017 | **-18.4%** |
| test_primary/surface_mse | — | 0.004106 (ep377 ckpt) | 0.0043 | — |
| full_test/volume_mse | — | 0.01654 (ep377 ckpt) | 0.0017 | — |

**Note:** Best surface and best volume occur at different epochs (377 vs 396). The checkpoint is selected by surface_mse, so test metrics come from the ep377 checkpoint — which has worse volume (0.01654 test) than the ep396 val-best volume (0.006235 val).

**Training:** 764 epochs completed (~5.5 hours, hit timeout). Plateau reached around epoch ~380. T_max=50 cosine cycling.

**Command:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --volume-loss-weight 30.0 \
  --agent megumi --wandb-name "megumi/af-volume-weight-30x" \
  --wandb-group af-volume-weight-sweep
```

**Peak memory:** ~34.4 GB (from nvidia-smi during training)
**W&B run:** `wupz6iuj`

### What happened

The 30x volume loss weight **did improve volume MSE by 18.4%** (0.00624 vs 0.00764 baseline), but at a **severe cost to surface MSE** (2.6x worse: 0.00126 vs 0.000482). The model is forced to allocate so much capacity to volume that surface prediction degrades heavily.

Key observations:
- **Volume improves but not dramatically.** 18% better is meaningful but the 4.5x gap to SpiderSolver (0.0017) barely narrows — we go from 4.5x to 3.7x.
- **Surface regresses far beyond the 9x headroom.** The PR noted 9x surface headroom over SpiderSolver. With 30x volume weight, surface MSE goes to 0.00126 — still beating SpiderSolver (0.0043) by 3.4x, but we burned most of that headroom.
- **Best surface and volume epochs diverge.** The Pareto front isn't achievable with a single checkpoint — surface peaks at ep377, volume at ep396.
- **Plateaued early.** No improvement after ~ep400 despite running to ep764. The 30x weight creates a loss landscape where the model saturates early.

### Suggested follow-ups

1. **Lower volume weights (3x, 10x)** — PRs #3101 and #3102 may find a better Pareto point. 10x might improve volume without destroying surface.
2. **Multi-objective checkpoint selection** — Select checkpoint by a weighted combination of surface + volume MSE, not just surface. The ep396 checkpoint has better volume but the current selection ignores it.
3. **Curriculum approach** — Start with 1x volume weight, gradually increase to 10-15x during training. This lets the model learn surface first (easier) then shift to volume.
4. **Separate surface/volume learning rate scaling** — Instead of loss weighting, use per-parameter-group LR to give volume-related parameters more gradient signal.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3134 because ADVISOR: vol-weight=30x is dead for AirfRANS. Surface=0.001264 (4.3x worse than baseline 0.000296), vol=0.006235 (3.1x worse than baseline 0.002039) (W&B: wupz6iuj). No Pareto improvement at any epoch. 30x massively overshoots — the optimizer saturates on volume gradients and both metrics collapse. The sweet spot was 10x (#3135, already merged). Adding to dead ends: vol-weight≥30x catastrophic for AF..
